fhir-server/src/main/java/au/csiro/pathling/fhirpath/encoding/SimpleCoding.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright © 2018-2020, Commonwealth Scientific and Industrial Research
+ * Organisation (CSIRO) ABN 41 687 119 230. Licensed under the CSIRO Open Source
+ * Software Licence Agreement.
+ */
+
+package au.csiro.pathling.fhirpath.encoding;
+
+import java.io.Serializable;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import org.hl7.fhir.r4.model.Coding;
+
+/**
+ * Implements a simplified version of {@link Coding} which is {@link Serializable} and includes an
+ * equality test.
+ *
+ * @author John Grimes
+ */
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class SimpleCoding implements Serializable {
+
+  private static final long serialVersionUID = 6509272647875353748L;
+
+  @Nullable
+  private String system;
+
+  @Nullable
+  private String code;
+
+  @Nullable
+  private String version;
+
+  /**
+   * @param system The code system URI for this coding
+   * @param code The code itself
+   */
+  public SimpleCoding(@Nullable final String system, @Nullable final String code) {
+    this(system, code, null);
+  }
+
+  /**
+   * @param coding A {@link Coding} object to copy values from
+   */
+  public SimpleCoding(@Nonnull final Coding coding) {
+    this.system = coding.getSystem();
+    this.code = coding.getCode();
+    this.version = coding.getVersion();
+  }
+
+  /**
+   * @return A new {@link Coding} containing the values from this SimpleCoding
+   */
+  @Nonnull
+  public Coding toCoding() {
+    return new Coding(system, code, version);
+  }
+
+  public boolean isNull() {
+    return system == null || code == null;
+  }
+
+  public boolean isNotNull() {
+    return !isNull();
+  }
+
+  /**
+   * Checks if the coding is versioned.
+   *
+   * @return true if the coding is versioned
+   */
+  public boolean isVersioned() {
+    return version != null;
+  }
+
+  /**
+   * Strips the version information from the coding
+   *
+   * @return unversioned coding
+   */
+  @Nonnull
+  public SimpleCoding toNonVersioned() {
+    return isVersioned()
+           ?
+           new SimpleCoding(this.system, this.code)
+           : this;
+  }
+}
```

Comment:
I think that this will also affect your implementation, as you use a Set here: https://github.com/aehrc/pathling/pull/134/files#diff-d7cc8ef2a328de3b080f47a81f6247b4R56

I think this might be falling back to the Lombok equality semantics.

Reply:
The equality semantics should match those in the `CodingsEqual` UDF.

Reply:
I think that if it does (match CodingsEqual semantics) it will also violate the java equals hashCode contract (for a equal versioned and unversioned codings)
I think we might need to add another method for this kind of equality. 
I think I compensate for this in the CodingSet implementation, that is if 
```
    boolean contains(@Nonnull final SimpleCoding c) {
      return allCodings.contains(c) || (c.isVersioned()
                                        ? allCodings.contains(c.toNonVersioned())
                                        : unversionedCodings.contains(c));
    }
```

Reply:
Yes, I hadn't thought of that. What you have done totally makes sense then.

The only thing is: if you don't want to rely upon `equals` / `hashCode`, should you be using a Set in `SubsumptionMapper`?

Reply:
I am not sure that I follow. As far as I understand Lombok @Data will generate equals/hashCode based on fields so the Set in SubsumptionMapper will work with this semantic and yes possibly send both versioned and unversioned copy of a coding to terminology server. But I think we did discuss that in the past and as far as I remember that is what it should do.

Reply:
Yes, I think you're right. Thanks.